File: src/test/scala/com/hpe/ossm/scluster/test/MyService.scala
package com.hpe.ossm.scluster.test

import akka.actor.{ActorRef, DeadLetter, Props}
import com.hpe.ossm.jcluster.messages.ServiceStatusEvents
import com.hpe.ossm.scluster.{ClusterNode, ServiceEntryActor}
import akka.pattern.Patterns.ask
import com.hpe.ossm.scluster.management.DeadLetterListener
import com.typesafe.config.Config

import scala.concurrent.duration._
import scala.collection.mutable
import scala.concurrent.ExecutionContext
import scala.util.{Failure, Success}

class MyService extends ServiceEntryActor("MyService", mutable.HashMap("CM" -> List.empty[ActorRef])) {
    implicit val ec: ExecutionContext = context.dispatcher

    override def preStart(): Unit = {
        super.preStart()
        timers.startPeriodicTimer("service", "Hello", 5.seconds)
    }

    override def postStop(): Unit = super.postStop()

    override def receive: Receive = super.receive.orElse {
        case s: String if ServiceStatusEvents.STATUS_AVAILABLE.equals(status) =>
            val ref = getService("CM")
            if (ref != null) {
                val reply = ask(ref, s, 5.seconds).mapTo[String]
                reply.onComplete {
                    case Success(r) =>println("[MyService Reply] " + r)
                    case Failure(e)=>println(s"Failed to ask. ${e.getMessage}")
                }
            }
        case s: Any => println(s"Ignored message ${s.toString}")
    }
}

class MyServiceApp extends ClusterNode("MyService", null) {}

object MyServiceNode {
    def main(args: Array[String]): Unit = {
        val t = new MyServiceApp()
        val ref = t.system.actorOf(Props(classOf[DeadLetterListener]), "deadLetterListener")
        t.system.eventStream.subscribe(ref, classOf[DeadLetter])
        t.system.actorOf(Props(classOf[MyService]), "myService")
    }
}

File: src/test/java/com/hpe/ossm/jcluster/test/MyService.java
package com.hpe.ossm.jcluster.test;


import akka.actor.ActorRef;
import akka.actor.DeadLetter;
import akka.actor.Props;
import akka.actor.ActorSystem;
import com.hpe.ossm.jcluster.ClusterNode;
import com.hpe.ossm.jcluster.ServiceEntryActor;
import com.hpe.ossm.jcluster.messages.ServiceStatusEvents;

import java.time.Duration;
import java.util.ArrayDeque;
import java.util.HashMap;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.CompletionException;

import static akka.pattern.Patterns.ask;

public class MyService extends ServiceEntryActor {

    @Override
    public void preStart() throws  Exception{
        //override the parameters
        serviceName = "MyService";
        dependServices = new HashMap<>();
        dependServices.put("CM", new ArrayDeque<>());
        super.preStart();
        timers().startPeriodicTimer("service","Hello",Duration.ofSeconds(30));
    }

    @Override
    public void postStop()throws  Exception {
        super.postStop();
    }

    @Override
    public Receive createReceive() {
        return super.createReceive()
                .orElse(
                        receiveBuilder()
                                .match(String.class, m -> {
                                    if(ServiceStatusEvents.STATUS_AVAILABLE.equals(status)){
                                        ActorRef ref = getService("CM");
                                        if (ref != null) {
                                            CompletableFuture<Object> future = ask(ref, m, Duration.ofSeconds(5)).toCompletableFuture();
                                            try {
                                                String reply = (String) future.join();
                                                System.out.println("[MyService Reply] " + reply);
                                            } catch (CompletionException e) {
                                                System.out.println("Failed to ask. " + e.getMessage());
                                            }
                                        }else{
                                            System.out.println("Ref to service CM is null");
                                        }
                                    }else{
                                        System.out.println("Service is not ready");
                                    }
                                })
                                .matchAny( o->System.out.println("Ignored message "+ o.toString()))
                                .build()
                );
    }
}

class MyServiceApp extends ClusterNode {

    public static void main(String[] args) {
        MyServiceApp t = new MyServiceApp();
        t.role="MyService";
        ActorSystem sys = t.getActorSystem(null);
        sys.actorOf(Props.create(MyService.class), "myService");
    }
}
